protocols/autonatv2/src/client/handler/dial_request.rs

```diff
@@ -0,0 +1,397 @@
+use futures::{
+    channel::{mpsc, oneshot},
+    AsyncRead, AsyncWrite, SinkExt, StreamExt,
+};
+use futures_bounded::FuturesSet;
+use libp2p_core::{
+    upgrade::{DeniedUpgrade, ReadyUpgrade},
+    Multiaddr,
+};
+
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    handler::{
+        ConnectionEvent, DialUpgradeError, FullyNegotiatedOutbound, OutboundUpgradeSend,
+        ProtocolsChange,
+    },
+    ConnectionHandler, ConnectionHandlerEvent, Stream, StreamProtocol, StreamUpgradeError,
+    SubstreamProtocol,
+};
+use std::{
+    collections::VecDeque,
+    convert::identity,
+    io,
+    iter::{once, repeat},
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use crate::request_response::Coder;
+use crate::{
+    generated::structs::{mod_DialResponse::ResponseStatus, DialStatus},
+    request_response::{
+        DialDataRequest, DialDataResponse, DialRequest, DialResponse, Request, Response,
+        DATA_FIELD_LEN_UPPER_BOUND, DATA_LEN_LOWER_BOUND, DATA_LEN_UPPER_BOUND,
+    },
+    REQUEST_PROTOCOL_NAME,
+};
+
+use super::{DEFAULT_TIMEOUT, MAX_CONCURRENT_REQUESTS};
+
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum InternalError {
+    #[error("io error")]
+    Io(#[from] io::Error),
+    #[error("invalid referenced address index: {index} (max number of addr: {max})")]
+    InvalidReferencedAddress { index: usize, max: usize },
+    #[error("data request too large: {len} (max: {max})")]
+    DataRequestTooLarge { len: usize, max: usize },
+    #[error("data request too small: {len} (min: {min})")]
+    DataRequestTooSmall { len: usize, min: usize },
+    #[error("timeout")]
+    Timeout(#[from] futures_bounded::Timeout),
+    #[error("server rejected dial request")]
+    ServerRejectedDialRequest,
+    #[error("server chose not to dial any provided address")]
+    ServerChoseNotToDialAnyAddress,
+    #[error("server ran into an internal error")]
+    InternalServer,
+    #[error("server did not respond correctly to dial request")]
+    InvalidResponse,
+    #[error("server was unable to connect to address: {addr:?}")]
+    UnableToConnectOnSelectedAddress { addr: Option<Multiaddr> },
+    #[error("server experienced failure during dial back on address: {addr:?}")]
+    FailureDuringDialBack { addr: Option<Multiaddr> },
+    #[error("error during substream upgrad")]
+    Substream(
+        #[from] StreamUpgradeError<<ReadyUpgrade<StreamProtocol> as OutboundUpgradeSend>::Error>,
+    ),
+}
+
+#[derive(Debug)]
+pub struct TestEnd {
+    pub(crate) dial_request: DialRequest,
+    pub(crate) reachable_addr: Multiaddr,
+}
+
+#[derive(Debug)]
+pub enum ToBehaviour {
+    TestCompleted(Result<TestEnd, super::Error>),
+    StatusUpdate(StatusUpdate),
+    PeerHasServerSupport,
+}
+
+#[derive(Debug)]
+pub struct StatusUpdate {
+    pub tested_addr: Option<Multiaddr>,
+    pub data_amount: usize,
+    pub server: PeerId,
+    pub result: Result<(), crate::client::handler::Error>,
```

Comment:
This `Error` should also be moved up to the `behaviour` module give that it is part of the public API.

Reply:
Done